salt/modules/arista_pyeapi.py

```diff
@@ -0,0 +1,577 @@
+    password: test
+
+In case the ``username`` and ``password`` are the same on any device you are
+targeting, the block above (besides other parameters specific to your
+environment you might need) should suffice to be able to execute commands from
+outside a ``pyeapi`` Proxy, e.g.:
+
+.. code-block:: bash
+
+    salt '*' pyeapi.send_commands 'show version' 'show interfaces'
+    salt '*' pyeapi.config 'ntp server 1.2.3.4'
+
+.. note::
+
+    Remember that the above applies only when not running in a ``pyeapi`` Proxy
+    Minion. If you want to use the :mod:`pyeapi Proxy <salt.proxy.arista_pyeapi>`,
+    please follow the documentation notes for a proper setup.
+'''
+from __future__ import absolute_import
+
+# Import python stdlib
+import logging
+
+# Import Salt libs
+from salt.ext import six
+from salt.exceptions import CommandExecutionError
+try:
+    from salt.utils.args import clean_kwargs
+    from salt.utils.files import mkstemp
+except ImportError:
+    from salt.utils import clean_kwargs
+    from salt.utils import mkstemp
+
+# Import third party libs
+try:
+    import pyeapi
+    HAS_PYEAPI = True
+except ImportError:
+    HAS_PYEAPI = False
+
+# -----------------------------------------------------------------------------
+# execution module properties
+# -----------------------------------------------------------------------------
+
+__proxyenabled__ = ['*']
+# Any Proxy Minion should be able to execute these
+
+__virtualname__ = 'pyeapi'
+# The Execution Module will be identified as ``pyeapi``
+
+# -----------------------------------------------------------------------------
+# globals
+# -----------------------------------------------------------------------------
+
+log = logging.getLogger(__name__)
+
+PYEAPI_INIT_KWARGS = [
+    'transport',
+    'host',
+    'username',
+    'password',
+    'enablepwd',
+    'port',
+    'timeout',
+    'return_node'
+]
+
+# -----------------------------------------------------------------------------
+# propery functions
+# -----------------------------------------------------------------------------
+
+
+def __virtual__():
+    '''
+    Execution module available only if Netmiko is installed.
+    '''
+    if not HAS_PYEAPI:
+        return False, 'The pyeapi execution module requires pyeapi library to be installed: ``pip install pyeapi``'
+    return __virtualname__
+
+# -----------------------------------------------------------------------------
+# helper functions
+# -----------------------------------------------------------------------------
+
+
+def _prepare_connection(**kwargs):
+    '''
+    Prepare the connection with the remote network device, and clean up the key
+    value pairs, removing the args used for the connection init.
+    '''
+    init_args = {}
+    fun_kwargs = {}
+    pyeapi_kwargs = __salt__['config.get']('pyeapi', {})
+    pyeapi_kwargs.update(kwargs)  # merge the CLI args with the opts/pillar
+    for karg, warg in six.iteritems(pyeapi_kwargs):
```

Comment:
We have nearly identical code over in netmiko_mod. Refactor for DRY?

Reply:
Apologies for late reply @cachedout. Yes, there's a little overlap between them. We could have the following function in the `utils`:

```python
def prepare_kwargs(all_kwargs, class_init_kwargs):
    for karg, warg in six.iteritems(all_kwargs):
        if karg not in class_init_kwargs:
            if warg is not None:
                fun_kwargs[karg] = warg
            continue
        if warg is not None:
            init_args[karg] = warg
    return init_args, fun_kwargs
```

Then the code in this execution module would become:

```python
def _prepare_connection(**kwargs):
    '''
    Prepare the connection with the remote network device, and clean up the key
    value pairs, removing the args used for the connection init.
    '''
    init_args = {}
    fun_kwargs = {}
    pyeapi_kwargs = __salt__['config.get']('pyeapi', {})
    pyeapi_kwargs.update(kwargs)  # merge the CLI args with the opts/pillar
    init_args, fun_kwargs = salt.utils.args.prepare_kwargs(pyeapi_kwargs, PYEAPI_INIT_KWARGS)
    if 'transport' not in init_args:
        init_args['transport'] = 'https'
    conn = pyeapi.client.connect(**init_args)
    node = pyeapi.client.Node(conn, enablepwd=init_args.get('enablepwd'))
    return node, fun_kwargs
```

I'm not disagreeing, just that I'm not sure if it's really worth it, particularly as it's harder to port a module in an older Salt version due to the utils dependency.
On the other hand, I have a couple of more modules pending in the queued to be submitted, that would have the same code as well.

What do you think would be the best? Thanks!

Reply:
If you use `__utils__['args.prepare_kwargs']` they should just have to drop the newer args module into `salt://_utils/` and sync them and it should be easily portable.

Reply:
Fantastic, I will do it like that then - will refactor and push tomorrow morning. Thanks guys!